File: gpio4j/src/main/java/org/overworld/gpio4j/serial/MessageType.java
package org.overworld.gpio4j.serial;

import java.util.Optional;
import java.util.stream.Stream;

import org.overworld.gpio4j.serial.wire.Analog;
import org.overworld.gpio4j.serial.wire.Config;
import org.overworld.gpio4j.serial.wire.Digital;
import org.overworld.gpio4j.serial.wire.Deconfig;
import org.overworld.gpio4j.serial.wire.Dump;
import org.overworld.gpio4j.serial.wire.Interrupt;
import org.overworld.gpio4j.serial.wire.Message;
import org.overworld.gpio4j.serial.wire.Notify;
import org.overworld.gpio4j.serial.wire.Read;
import org.overworld.gpio4j.serial.wire.Write;

import lombok.Getter;

public enum MessageType {
	READ(1), WRITE(2), CONFIG(3), DECONFIG(4), DUMP(5), DIGITAL(6), ANALOG(7), INTERRUPT(8), NOTIFY(9);

	@Getter
	private int id;

	private MessageType(int type) {
		this.id = type;
	}
	
	public static Optional<MessageType> forId(int id) {
		return Stream.of(MessageType.values()).filter(mt-> id == mt.getId()).findFirst();
	}
	
	public Message newInstance() {
		return switch (this) {
			case CONFIG -> new Config();
			case DIGITAL -> new Digital();
			case DECONFIG -> new Deconfig();
			case DUMP -> new Dump();
			case INTERRUPT -> new Interrupt();
			case READ -> new Read();
			case WRITE -> new Write();
		    case ANALOG -> new Analog();
		    case NOTIFY -> new Notify();
		};
	}
}

File: BoardDriver/src/MessageType.cpp
#include "MessageType.h"

class MessageTypeHelper {
public:
  static uint8_t getId(MessageType type) {
    return static_cast<uint8_t>(type);
  }

  static MessageType forId(uint8_t id) { // @suppress("No return")
    for (int i = static_cast<uint8_t>(MessageType::READ); i <= static_cast<uint8_t>(MessageType::NOTIFY); ++i) {
      if (id == i) {
        return static_cast<MessageType>(i);
      }
    }
  }

  static std::unique_ptr<Message> newInstance(MessageType type) { // @suppress("No return")
    switch (type) {
      case MessageType::CONFIG:
        return Config();
      case MessageType::DIGITAL:
        return Digital();
      case MessageType::DECONFIG:
        return Deconfig();
      case MessageType::DUMP:
        return Dump();
      case MessageType::INTERRUPT:
        return Interrupt();
      case MessageType::READ:
        return Read();
      case MessageType::WRITE:
        return Write();
      case MessageType::ANALOG:
        return Analog();
      case MessageType::NOTIFY:
        return Notify();
    }
  }
};
